xdsl/passes.py

```diff
@@ -1,20 +1,123 @@
 from abc import ABC, abstractmethod
+from dataclasses import dataclass, Field
+from types import UnionType, NoneType
+
+from xdsl.utils.hints import isa
 from xdsl.dialects import builtin
 from xdsl.ir import MLContext
-from typing import ClassVar
+from typing import ClassVar, TypeVar, get_origin, Any, Union, get_args
+from xdsl.utils.parse_pipeline import (
+    PipelinePassSpec,
+    PassArgListType,
+    PassArgElementType,
+)
+
+_T = TypeVar("_T", bound="ModulePass")
 
 
+@dataclass
 class ModulePass(ABC):
     """
     A Pass is a named rewrite pass over an IR module.
 
     All passes are expected to leave the IR in a valid state after application.
     That is, the IR verifies. In turn, all passes can expect the IR they are
-    applied to to be in a valid state.
+    applied to be in a valid state.
     """
 
     name: ClassVar[str]
 
     @abstractmethod
     def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
         ...
+
+    @classmethod
+    def from_pass_spec(cls: type[_T], spec: PipelinePassSpec) -> _T:
+        """
+        This method takes a PipelinePassSpec, does type checking on the
+        arguments, and then instantiates an instance of the ModulePass
+        from the spec.
+        """
+        # some sanity checks here
+        assert issubclass(cls, ModulePass), f"{cls} must be subclass of ModulePass"
+        assert hasattr(cls, "__dataclass_fields__"), f"{cls} must be a dataclass"
+        assert spec.name == cls.name, "Wrong pass name provided"
```

Comment:
I made these asserts because they should never happen under "normal" operation, if that makes sense? I can change them to value errors, just thought that was the procedure in the codebase.